docs: added cart conceptual guide (#3872)
medusajs/medusa: docs/content/modules/carts-and-checkout/cart.md

---
description: 'Learn about the Cart entity, its relation to other entities, and how the cart completion process is implemented.'
---

# Cart Architecture Overview

In this document, you’ll learn about the Cart entity, its relation to other entities, and how the cart completion strategy is implemented.

## Overview

> The cart is a virtual shopping cart that customers use to choose products they want to eventually purchase. They can set the quantity, add discounts, and more before they proceed to the checkout.

Reviewer: ```suggestion
The cart allows customers to go from browsing products to completing an order. Many actions can be performed on a cart, like adding line items, applying discounts, creating shipping methods, and more before eventually completing it and creating an Order.
```

We could link to Order in the last sentence?

Author: I'm actually currently working on the Order conceptual guide, so I'll do the linking between them then 👍🏻

> In Medusa, a cart isn’t only used to check out and place an order. It’s also used to create swaps.

Reviewer: I would remove this and let the swaps guide explain it. I dread that only adding a single sentence can cause more harm (read confusion) than good.
```suggestion
```

---

## Cart Entity Overview

A cart is represented by the `Cart` entity. Some of the `Cart` entity’s attributes include:

- `email`: The email the cart is associated with.
- `type`: A string indicating what the cart is used for. Its value can be:
  - `default` if the cart is used to place an order.
  - `swap` if the cart is used to create and register a swap
  - `draft_order` if the cart is used to create and complete a draft order.
  - `payment_link` if the cart is used for a payment link.
  - `claim` if the cart is used to create a claim.
- `completed_at`: the date the cart was completed. A completed cart means that it has been used for its main purpose. For example, if the cart is used to place an order, then a completed cart means that the order was placed.
- `payment_authorized_at`: the date a payment was authorized.

> There are other important attributes discussed in later sections. You can also check out the [full Cart entity in the entities reference](../../references/entities/classes/Cart.md).

Reviewer: ```suggestion
There are other important attributes discussed in later sections. Check out the [full Cart entity in the entities reference](../../references/entities/classes/Cart.md).
```

---

## Cart Totals Calculation

> By default, the cart entity doesn’t hold any details regarding cart totals. These are created and added to a cart object using the `CartService`'s [decorateTotals method](../../references/services/classes/CartService.md#decoratetotals). This method is used by endpoints and services that require retrieving the cart’s totals.

Reviewer: ```suggestion
By default, the cart entity doesn’t hold any details regarding the totals. These are computed and added to the cart entity using the `CartService`'s [decorateTotals method](../../references/services/classes/CartService.md#decoratetotals). There's also a dedicated method in the `CartService`, [retrieveWithTotals](../../references/services/classes/CartService.md#retrieveWithTotals), attaching the totals, so that you don't have to worry about it. It is recommended to use this method by default when you need to retrieve the cart. 
```

> The cart’s totals are created based on the context of the cart. This includes the selected region, whether tax-inclusive pricing is enabled, the chosen shipping methods, and more.

Reviewer: ```suggestion
The cart’s totals are calculated based on the content and context of the cart. This includes the selected region, whether tax-inclusive pricing is enabled, the chosen shipping methods, and more.
```

The calculated cart’s totals include:

- `shipping_total`: The total of the chosen shipping methods, with taxes.
- `shipping_tax_total`: The applied taxes on the shipping total.
- `discount_total`: The total of the applied discounts.
- `raw_discount_total`: The total of the applied discounts without rounding.
- `item_tax_total`: The total applied taxes on the cart’s items.
- `tax_total`: The total taxes applied (the sum of `shipping_tax_total` and `item_tax_total`).
- `gift_card_total`: The total gift card amount applied on the cart. If there are any taxes applied on the gift cards, they’re deducted from the total.
- `gift_card_tax_total`: The total taxes applied on the cart’s gift cards.
- `subtotal`: The total of the items without taxes or discounts.
- `total`: The overall total of the cart.

:::note

If you have tax-inclusive pricing enabled, you can learn about other available total fields [here](../taxes/inclusive-pricing.md#retrieving-tax-amounts).

:::

> The cart’s totals are retrieved by default in all the [cart’s store APIs](https://docs.medusajs.com/api/store#tag/Carts). You can also retrieve a cart with its totals within your code using the `CartService`'s [retrieveWithTotals method](../../references/services/classes/CartService.md#retrievewithtotals).

Reviewer: I've added it to a section above instead.
```suggestion
The cart’s totals are retrieved by default in all the [cart’s store APIs](https://docs.medusajs.com/api/store#tag/Carts).
```

---

## Cart Completion

The cart completion functionality is implemented inside the strategy `cartCompletionStrategy`. This allows you to customize how the process is implemented in your store.

> The cart completion process starts when the checkout process is completed by sending a request to the [Complete Cart endpoint](https://docs.medusajs.com/api/store#tag/Carts/operation/PostCartsCartComplete).

Reviewer: I would reduce this to be centered around only completing the cart. We don't know what the checkout process looks like. With our flexible tooling, this can be a wide variety of flows, and there's no reason to focus on conventional e-commerce here.
```suggestion
You initiate the cart completion process by sending a request to the [Complete Cart endpoint](https://docs.medusajs.com/api/store#tag/Carts/operation/PostCartsCartComplete).
```

Reviewer: This comment is not directly related to this paragraph but more a general thing to have in mind for the guide. I hope it makes sense.

> ### Idempotency Key

Reviewer: praise: I like it!

An Idempotency Key is a unique key associated with a cart. It is generated when the cart completion process is started. The idempotency key is stored in the Cart entity under the attribute `idempotency_key`.

That Idempotency Key is then set in the header under the `Idempotency-Key` response header field along with the header field `Access-Control-Expose-Headers` set to `Idempotency-Key`.

The backend wraps essential parts of the checkout completion in its own step and stores the current step (known as recovery point) of checkout with its associated Idempotency Key.

Then, if the request is interrupted for any reason or an error occurs, the client can retry completing the checkout using the Idempotency Key, and the flow will continue from the last stored step.

> ### Cart Completion Process

Reviewer: question: Are we including a separate page on overriding the strategy? 

That would be immensely valuable.

Author: It's not available in the moment but yes in the future we will

> ![Cart Completion Flowchart](https://res.cloudinary.com/dza7lstvk/image/upload/v1681828705/Medusa%20Docs/Diagrams/cart-completion_atqkhg.jpg)

Reviewer: praise: This is also very nice!

The process is implemented as follows:

1. When the idempotency key’s recovery point is set to `started`, the tax lines are created for the items in the cart. This is done using the `CartService`'s [createTaxLines method](../../references/services/classes/CartService.md#createtaxlines). If that is completed with no errors, the recovery point is set to `tax_lines_created` and the process continues.
2. When the idempotency key’s recovery point is set to `tax_lines_created`, the payment is authorized using the `CartService`'s method [authorizePayment](../../references/services/classes/CartService.md#authorizepayment). If the payment requires more action or is pending authorization, then the tax lines that were created in the previous steps are deleted and the cart completion process is terminated. Once the payment is authorized, the process can be restarted.

> 3. When the idempotency key’s recovery point is set to `payment_authorized`, tax lines are created again the same way as the first step. Then, the inventory of each of the items in the cart is confirmed using the `ProductVariantInventoryService`'s method [confirmInventory](../../references/services/classes/ProductVariantInventoryService.md#confirminventory). If an item’s quantity is available in the inventory, the quantity is reserved using the `ProductVariantInventoryService`'s method [reserveQuantity](../../references/services/classes/ProductVariantInventoryService.md#reservequantity). If any item has insufficient inventory, any item reservations that were created are deleted, the payment is canceled, and an error is thrown, terminating the cart completion process. If all item quantities are confirmed to be available:

Reviewer: ```suggestion
3. When the idempotency key’s recovery point is set to `payment_authorized`, tax lines are created again the same way as the first step. Then, the inventory of each of the items in the cart is confirmed using the `ProductVariantInventoryService`'s method [confirmInventory](../../references/services/classes/ProductVariantInventoryService.md#confirminventory). If an item is in stock, the quantity is reserved using the `ProductVariantInventoryService`'s method [reserveQuantity](../../references/services/classes/ProductVariantInventoryService.md#reservequantity). If an item is out of stock, any item reservations that were created are deleted, the payment is canceled, and an error is thrown, terminating the cart completion process. If all item quantities are confirmed to be available:
```
